// cozy-client/dist/queries/dsl.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.QueryDefinition = exports.MutationTypes = exports.Mutations = exports.getDoctypeFromOperation = exports.uploadFile = exports.removeReferencedBy = exports.addReferencedBy = exports.removeReferencesTo = exports.addReferencesTo = exports.deleteDocument = exports.updateDocument = exports.createDocument = exports.Q = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var isArray = require('lodash/isArray');
/**
 * typedef QueryDefinition
 */

/**
 * Chainable API to create query definitions to retrieve documents
 * from a Cozy. `QueryDefinition`s are sent to links.
 */


var QueryDefinition =
/*#__PURE__*/
function () {
  /**
   * @class
   * @param {object} options Initial options for the query definition
   * @param {string} options.doctype - The doctype of the doc.
   * @param {string} options.id - The id of the doc.
   * @param {Array} options.ids - The ids of the docs.
   * @param {object} options.selector - The selector to query the docs.
   * @param {Array} options.fields - The fields to return.
   * @param {Array} options.indexedFields - The fields to index.
   * @param {Array} options.sort - The sorting params.
   * @param {string} options.includes - The docs to include.
   * @param {string} options.referenced - The referenced document.
   * @param {number} options.limit - The document's limit to return.
   * @param {number} options.skip - The number of docs to skip.
   * @param {number} options.cursor - The cursor to paginate views.
   * @param {number} options.bookmark - The bookmark to paginate mango queries.
   */
  function QueryDefinition() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck2.default)(this, QueryDefinition);
    this.doctype = options.doctype;
    this.id = options.id;
    this.ids = options.ids;
    this.selector = options.selector;
    this.fields = options.fields;
    this.indexedFields = options.indexedFields;
    this.sort = options.sort;
    this.includes = options.includes;
    this.referenced = options.referenced;
    this.limit = options.limit;
    this.skip = options.skip;
    this.cursor = options.cursor;
    this.bookmark = options.bookmark;
  }
  /**
   * Query a single document on its id.
   *
   * @param {string} id   The document id.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */


  (0, _createClass2.default)(QueryDefinition, [{
    key: "getById",
    value: function getById(id) {
      if (!id) {
        throw new Error('getById called with undefined id');
      }

      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        id: id
      }));
    }
    /**
     * Query several documents on their ids.
     *
     * @param {Array} ids   The documents ids.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "getByIds",
    value: function getByIds(ids) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        ids: ids
      }));
    }
    /**
     * Query documents with a [mango selector](http://docs.couchdb.org/en/latest/api/database/find.html#find-selectors).
     * Each field passed in the selector will be indexed, except if the indexField option is used.
     *
     * @param {object} selector   The Mango selector.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "where",
    value: function where(selector) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        selector: selector
      }));
    }
    /**
     * Specify which fields of each object should be returned. If it is omitted, the entire object is returned.
     *
     * @param {Array} fields The fields to return.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "select",
    value: function select(fields) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        fields: fields
      }));
    }
    /**
     * Specify which fields should be indexed. This prevent the automatic indexing of the mango fields.
     *
     * @param {Array} indexedFields The fields to index.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "indexFields",
    value: function indexFields(indexedFields) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        indexedFields: indexedFields
      }));
    }
    /**
     * Specify how to sort documents, following the [sort syntax](http://docs.couchdb.org/en/latest/api/database/find.html#find-sort)
     *
     * @param {Array} sort The list of field name and direction pairs.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "sortBy",
    value: function sortBy(sort) {
      if (!isArray(sort)) {
        throw new Error("Invalid sort, should be an array ([{ label: \"desc\"}, { name: \"asc\"}]), you passed ".concat(JSON.stringify(sort), "."));
      }

      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        sort: sort
      }));
    }
    /**
     * Includes documents having a relationships with the ones queried.
     * For example, query albums including the photos.
     *
     * @param {Array} includes The documents to include.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "include",
    value: function include(includes) {
      if (!Array.isArray(includes)) {
        throw new Error('include() takes an array of relationship names');
      }

      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        includes: includes
      }));
    }
    /**
     * Maximum number of documents returned, useful for pagination. Default is 100.
     *
     * @param {number} limit The document's limit.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "limitBy",
    value: function limitBy(limit) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        limit: limit
      }));
    }
  }, {
    key: "UNSAFE_noLimit",
    value: function UNSAFE_noLimit() {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        limit: null
      }));
    }
    /**
     * Skip the first ‘n’ documents, where ‘n’ is the value specified.
     *
     * Beware, this [performs badly](http://docs.couchdb.org/en/stable/ddocs/views/pagination.html#paging-alternate-method) on view's index.
     *  Prefer cursor-based pagination in such situation.
     *
     * @param {number} skip The number of documents to skip.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "offset",
    value: function offset(skip) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        bookmark: undefined,
        cursor: undefined,
        skip: skip
      }));
    }
    /**
     * Use [cursor-based](https://docs.cozy.io/en/cozy-stack/jsonapi/#pagination) pagination.
     * *Warning*: this is only useful for views.
     * The cursor is a [startkey, startkey_docid] array, where startkey is the view's key,
     * e.g. ["io.cozy.photos.albums", "album-id"] and startkey_docid is the id of
     * the starting document of the query, e.g. "file-id".
     * Use the last docid of each query as startkey_docid to paginate or leave blank for the first query.
     *
     * @param {Array} cursor The cursor for pagination.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "offsetCursor",
    value: function offsetCursor(cursor) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        bookmark: undefined,
        skip: undefined,
        cursor: cursor
      }));
    }
    /**
     * Use [bookmark](https://docs.couchdb.org/en/2.2.0/api/database/find.html#pagination) pagination.
     * Note this only applies for mango-queries (not views) and is way more efficient than skip pagination.
     * The bookmark is a string returned by the _find response and can be seen as a pointer in
     * the index for the next query.
     *
     * @param {string} bookmark The bookmark to continue a previous paginated query.
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "offsetBookmark",
    value: function offsetBookmark(bookmark) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        skip: undefined,
        cursor: undefined,
        bookmark: bookmark
      }));
    }
    /**
     * Use the [file reference system](https://docs.cozy.io/en/cozy-stack/references-docs-in-vfs/)
     *
     * @param {object} document The reference document
     * @returns {QueryDefinition}  The QueryDefinition object.
     */

  }, {
    key: "referencedBy",
    value: function referencedBy(document) {
      return new QueryDefinition((0, _objectSpread2.default)({}, this.toDefinition(), {
        referenced: document
      }));
    }
  }, {
    key: "toDefinition",
    value: function toDefinition() {
      return {
        doctype: this.doctype,
        id: this.id,
        ids: this.ids,
        selector: this.selector,
        fields: this.fields,
        indexedFields: this.indexedFields,
        sort: this.sort,
        includes: this.includes,
        referenced: this.referenced,
        limit: this.limit,
        skip: this.skip,
        cursor: this.cursor,
        bookmark: this.bookmark
      };
    }
  }]);
  return QueryDefinition;
}();
/**
 * Helper to create a QueryDefinition. Recommended way to create
 * query definitions.
 *
 * @example
 * ```
 * import { Q } from 'cozy-client'
 *
 * const qDef = Q('io.cozy.todos').where({ _id: '1234' })
 * ```
 */


exports.QueryDefinition = QueryDefinition;

var Q = function Q(doctype) {
  return new QueryDefinition({
    doctype: doctype
  });
}; // Mutations


exports.Q = Q;
var CREATE_DOCUMENT = 'CREATE_DOCUMENT';
var UPDATE_DOCUMENT = 'UPDATE_DOCUMENT';
var DELETE_DOCUMENT = 'DELETE_DOCUMENT';
var ADD_REFERENCES_TO = 'ADD_REFERENCES_TO';
var REMOVE_REFERENCES_TO = 'REMOVE_REFERENCES_TO';
var ADD_REFERENCED_BY = 'ADD_REFERENCED_BY';
var REMOVE_REFERENCED_BY = 'REMOVE_REFERENCED_BY';
var UPLOAD_FILE = 'UPLOAD_FILE';

var createDocument = function createDocument(document) {
  return {
    mutationType: MutationTypes.CREATE_DOCUMENT,
    document: document
  };
};

exports.createDocument = createDocument;

var updateDocument = function updateDocument(document) {
  return {
    mutationType: MutationTypes.UPDATE_DOCUMENT,
    document: document
  };
};

exports.updateDocument = updateDocument;

var deleteDocument = function deleteDocument(document) {
  return {
    mutationType: MutationTypes.DELETE_DOCUMENT,
    document: document
  };
};

exports.deleteDocument = deleteDocument;

var addReferencesTo = function addReferencesTo(document, referencedDocuments) {
  return {
    mutationType: MutationTypes.ADD_REFERENCES_TO,
    referencedDocuments: referencedDocuments,
    document: document
  };
};

exports.addReferencesTo = addReferencesTo;

var removeReferencesTo = function removeReferencesTo(document, referencedDocuments) {
  return {
    mutationType: MutationTypes.REMOVE_REFERENCES_TO,
    referencedDocuments: referencedDocuments,
    document: document
  };
};

exports.removeReferencesTo = removeReferencesTo;

var addReferencedBy = function addReferencedBy(document, referencedDocuments) {
  return {
    mutationType: MutationTypes.ADD_REFERENCED_BY,
    referencedDocuments: referencedDocuments,
    document: document
  };
};

exports.addReferencedBy = addReferencedBy;

var removeReferencedBy = function removeReferencedBy(document, referencedDocuments) {
  return {
    mutationType: MutationTypes.REMOVE_REFERENCED_BY,
    referencedDocuments: referencedDocuments,
    document: document
  };
};

exports.removeReferencedBy = removeReferencedBy;

var uploadFile = function uploadFile(file, dirPath) {
  return {
    mutationType: MutationTypes.UPLOAD_FILE,
    file: file,
    dirPath: dirPath
  };
};

exports.uploadFile = uploadFile;

var getDoctypeFromOperation = function getDoctypeFromOperation(operation) {
  if (operation.mutationType) {
    var type = operation.mutationType;

    switch (type) {
      case CREATE_DOCUMENT:
        return operation.document._type;

      case UPDATE_DOCUMENT:
        return operation.document._type;

      case DELETE_DOCUMENT:
        return operation.document._type;

      case ADD_REFERENCES_TO:
        throw new Error('Not implemented');

      case UPLOAD_FILE:
        throw new Error('Not implemented');

      default:
        throw new Error("Unknown mutationType ".concat(type));
    }
  } else {
    return operation.doctype;
  }
};

exports.getDoctypeFromOperation = getDoctypeFromOperation;
var Mutations = {
  createDocument: createDocument,
  updateDocument: updateDocument,
  deleteDocument: deleteDocument,
  addReferencesTo: addReferencesTo,
  removeReferencesTo: removeReferencesTo,
  addReferencedBy: addReferencedBy,
  removeReferencedBy: removeReferencedBy,
  uploadFile: uploadFile
};
exports.Mutations = Mutations;
var MutationTypes = {
  CREATE_DOCUMENT: CREATE_DOCUMENT,
  UPDATE_DOCUMENT: UPDATE_DOCUMENT,
  DELETE_DOCUMENT: DELETE_DOCUMENT,
  ADD_REFERENCES_TO: ADD_REFERENCES_TO,
  REMOVE_REFERENCES_TO: REMOVE_REFERENCES_TO,
  ADD_REFERENCED_BY: ADD_REFERENCED_BY,
  REMOVE_REFERENCED_BY: REMOVE_REFERENCED_BY,
  UPLOAD_FILE: UPLOAD_FILE
};
exports.MutationTypes = MutationTypes;
