// cozy-stack-client/dist/AppCollection.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.normalizeApp = exports.APPS_DOCTYPE = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _DocumentCollection2 = _interopRequireWildcard(require("./DocumentCollection"));

var _errors = require("./errors");

var APPS_DOCTYPE = 'io.cozy.apps';
exports.APPS_DOCTYPE = APPS_DOCTYPE;

var normalizeApp = function normalizeApp(app, doctype) {
  return (0, _objectSpread2.default)({}, app, (0, _DocumentCollection2.normalizeDoc)(app, doctype), app.attributes);
};
/**
 * Extends `DocumentCollection` API along with specific methods for `io.cozy.apps`.
 */


exports.normalizeApp = normalizeApp;

var AppCollection =
/*#__PURE__*/
function (_DocumentCollection) {
  (0, _inherits2.default)(AppCollection, _DocumentCollection);

  function AppCollection(stackClient) {
    var _this;

    (0, _classCallCheck2.default)(this, AppCollection);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(AppCollection).call(this, APPS_DOCTYPE, stackClient));
    _this.endpoint = '/apps/';
    return _this;
  }
  /**
   * Lists all apps, without filters.
   *
   * The returned documents are not paginated by the stack.
   *
   * @returns {{data, meta, skip, next}} The JSON API conformant response.
   * @throws {FetchError}
   */


  (0, _createClass2.default)(AppCollection, [{
    key: "all",
    value: function () {
      var _all = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee() {
        var _this2 = this;

        var resp;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.stackClient.fetchJSON('GET', this.endpoint);

              case 2:
                resp = _context.sent;
                return _context.abrupt("return", {
                  data: resp.data.map(function (app) {
                    return normalizeApp(app, _this2.doctype);
                  }),
                  meta: {
                    count: resp.meta.count
                  },
                  skip: 0,
                  next: false
                });

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function all() {
        return _all.apply(this, arguments);
      };
    }()
  }, {
    key: "create",
    value: function () {
      var _create = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2() {
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                throw new Error('create() method is not available for applications');

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      return function create() {
        return _create.apply(this, arguments);
      };
    }()
  }, {
    key: "update",
    value: function () {
      var _update = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3() {
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                throw new Error('update() method is not available for applications');

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3);
      }));

      return function update() {
        return _update.apply(this, arguments);
      };
    }()
  }, {
    key: "destroy",
    value: function () {
      var _destroy = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee4() {
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                throw new Error('destroy() method is not available for applications');

              case 1:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4);
      }));

      return function destroy() {
        return _destroy.apply(this, arguments);
      };
    }()
  }]);
  return AppCollection;
}(_DocumentCollection2.default);

AppCollection.normalizeDoctype = _DocumentCollection2.default.normalizeDoctypeJsonApi;
var _default = AppCollection;
exports.default = _default;
