// cozy-stack-client/dist/memoize.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorReturned = exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));

var ErrorReturned =
/*#__PURE__*/
function (_String) {
  (0, _inherits2.default)(ErrorReturned, _String);

  function ErrorReturned() {
    (0, _classCallCheck2.default)(this, ErrorReturned);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ErrorReturned).apply(this, arguments));
  }

  return ErrorReturned;
}((0, _wrapNativeSuper2.default)(String));

exports.ErrorReturned = ErrorReturned;

/**
 * Delete outdated results from cache
 */
var garbageCollect = function garbageCollect(cache, maxDuration) {
  var now = Date.now();

  for (var _i = 0, _Object$keys = Object.keys(cache); _i < _Object$keys.length; _i++) {
    var key = _Object$keys[_i];
    var delta = now - cache[key].date;

    if (delta > maxDuration) {
      delete cache[key];
    }
  }
};

var isPromise = function isPromise(maybePromise) {
  return typeof maybePromise === 'object' && typeof maybePromise.then === 'function';
};
/**
 * Memoize with maxDuration and custom key
 */


var memoize = function memoize(fn, options) {
  var cache = {};
  return function () {
    var key = options.key.apply(null, arguments);
    garbageCollect(cache, options.maxDuration);
    var existing = cache[key];

    if (existing) {
      return existing.result;
    } else {
      var result = fn.apply(this, arguments);
      cache[key] = {
        result: result,
        date: Date.now()
      };
      /**
       * If the result is a promise and this promise
       * failed or resolved with a specific error (aka ErrorReturned),
       * let's remove the result from the cache since we don't want to
       * memoize error
       */

      if (isPromise(result)) {
        result.then(function (v) {
          if (v instanceof ErrorReturned) {
            delete cache[key];
          }
        }).catch(function (e) {
          delete cache[key];
        });
      }

      return result;
    }
  };
};

var _default = memoize;
exports.default = _default;
