// cozy-stack-client/dist/OAuthClient.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _CozyStackClient2 = _interopRequireDefault(require("./CozyStackClient"));

var _AccessToken = _interopRequireDefault(require("./AccessToken"));

var _logDeprecate = _interopRequireDefault(require("./logDeprecate"));

var defaultoauthOptions = {
  clientID: '',
  clientName: '',
  clientKind: '',
  clientSecret: '',
  clientURI: '',
  registrationAccessToken: '',
  redirectURI: '',
  softwareID: '',
  softwareVersion: '',
  logoURI: '',
  policyURI: '',
  notificationPlatform: '',
  notificationDeviceToken: ''
};
/**
 * Specialized `CozyStackClient` for mobile, implementing stack registration
 * through OAuth.
 */

var OAuthClient =
/*#__PURE__*/
function (_CozyStackClient) {
  (0, _inherits2.default)(OAuthClient, _CozyStackClient);

  function OAuthClient(_ref) {
    var _this;

    var oauth = _ref.oauth,
        _ref$scope = _ref.scope,
        scope = _ref$scope === void 0 ? [] : _ref$scope,
        onTokenRefresh = _ref.onTokenRefresh,
        options = (0, _objectWithoutProperties2.default)(_ref, ["oauth", "scope", "onTokenRefresh"]);
    (0, _classCallCheck2.default)(this, OAuthClient);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(OAuthClient).call(this, options));

    _this.setOAuthOptions((0, _objectSpread2.default)({}, defaultoauthOptions, oauth));

    if (oauth.token) {
      _this.setToken(oauth.token);
    }

    _this.scope = scope;
    _this.onTokenRefresh = onTokenRefresh;
    return _this;
  }
  /**
   * Checks if the client has his registration information from the server
   *
   * @returns {boolean} true if registered, false otherwise
   * @private
   */


  (0, _createClass2.default)(OAuthClient, [{
    key: "isRegistered",
    value: function isRegistered() {
      return this.oauthOptions.clientID !== '';
    }
    /**
     * Converts a camel-cased data set to snake case, suitable for sending to the OAuth server
     *
     * @param   {object} data Initial data
     * @returns {object} Formatted data
     * @private
     */

  }, {
    key: "snakeCaseOAuthData",
    value: function snakeCaseOAuthData(data) {
      var mappedFields = {
        softwareID: 'software_id',
        softwareVersion: 'software_version',
        clientID: 'client_id',
        clientName: 'client_name',
        clientKind: 'client_kind',
        clientURI: 'client_uri',
        logoURI: 'logo_uri',
        policyURI: 'policy_uri',
        notificationPlatform: 'notification_platform',
        notificationDeviceToken: 'notification_device_token',
        redirectURI: 'redirect_uris'
      };
      var result = {};
      Object.keys(data).forEach(function (fieldName) {
        var key = mappedFields[fieldName] || fieldName;
        var value = data[fieldName];
        result[key] = value;
      }); // special case: turn redirect_uris into an array

      if (result['redirect_uris'] && result['redirect_uris'] instanceof Array === false) result['redirect_uris'] = [result['redirect_uris']];
      return result;
    }
    /**
     * Converts a snake-cased data set to camel case, suitable for internal use
     *
     * @param   {object} data Initial data
     * @returns {object} Formatted data
     * @private
     */

  }, {
    key: "camelCaseOAuthData",
    value: function camelCaseOAuthData(data) {
      var mappedFields = {
        client_id: 'clientID',
        client_name: 'clientName',
        client_secret: 'clientSecret',
        registration_access_token: 'registrationAccessToken',
        software_id: 'softwareID',
        redirect_uris: 'redirectURI'
      };
      var result = {};
      Object.keys(data).forEach(function (fieldName) {
        var key = mappedFields[fieldName] || fieldName;
        var value = data[fieldName];
        result[key] = value;
      });
      return result;
    }
    /** Performs the HTTP call to register the client to the server */

  }, {
    key: "doRegistration",
    value: function doRegistration() {
      return this.fetchJSON('POST', '/auth/register', this.snakeCaseOAuthData({
        redirectURI: this.oauthOptions.redirectURI,
        clientName: this.oauthOptions.clientName,
        softwareID: this.oauthOptions.softwareID,
        clientKind: this.oauthOptions.clientKind,
        clientURI: this.oauthOptions.clientURI,
        logoURI: this.oauthOptions.logoURI,
        policyURI: this.oauthOptions.policyURI,
        softwareVersion: this.oauthOptions.softwareVersion,
        notificationPlatform: this.oauthOptions.notificationPlatform,
        notificationDeviceToken: this.oauthOptions.notificationDeviceToken
      }));
    }
    /**
     * Registers the currenly configured client with the OAuth server and
     * sets internal information from the server response
     *
     * @throws {Error} When the client is already registered
     * @returns {Promise} A promise that resolves with a complete list of client information, including client ID and client secret.
     */

  }, {
    key: "register",
    value: function () {
      var _register = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee() {
        var mandatoryFields, fields, missingMandatoryFields, data;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this.isRegistered()) {
                  _context.next = 2;
                  break;
                }

                throw new Error('Client already registered');

              case 2:
                mandatoryFields = ['redirectURI'];
                fields = Object.keys(this.oauthOptions);
                missingMandatoryFields = mandatoryFields.filter(function (fieldName) {
                  return fields[fieldName];
                });

                if (!(missingMandatoryFields.length > 0)) {
                  _context.next = 7;
                  break;
                }

                throw new Error("Can't register client : missing ".concat(missingMandatoryFields, " fields"));

              case 7:
                _context.next = 9;
                return this.doRegistration();

              case 9:
                data = _context.sent;
                this.setOAuthOptions((0, _objectSpread2.default)({}, this.oauthOptions, {
                  client_id: data.client_id,
                  client_name: data.client_name,
                  client_secret: data.client_secret,
                  registration_access_token: data.registration_access_token,
                  software_id: data.software_id
                }));
                return _context.abrupt("return", this.oauthOptions);

              case 12:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function register() {
        return _register.apply(this, arguments);
      };
    }()
    /**
     * Unregisters the currenly configured client with the OAuth server.
     *
     * @throws {NotRegisteredException} When the client doesn't have it's registration information
     * @returns {Promise}
     */

  }, {
    key: "unregister",
    value: function () {
      var _unregister = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2() {
        var clientID;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (this.isRegistered()) {
                  _context2.next = 2;
                  break;
                }

                throw new NotRegisteredException();

              case 2:
                clientID = this.oauthOptions.clientID;
                this.oauthOptions.clientID = '';
                return _context2.abrupt("return", this.fetchJSON('DELETE', "/auth/register/".concat(clientID), null, {
                  headers: {
                    Authorization: this.registrationAccessTokenToAuthHeader()
                  }
                }));

              case 5:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function unregister() {
        return _unregister.apply(this, arguments);
      };
    }()
    /**
     * Fetches the complete set of client information from the server after it has been registered.
     *
     * @throws {NotRegisteredException} When the client doesn't have it's registration information
     * @returns {Promise}
     */

  }, {
    key: "fetchInformation",
    value: function () {
      var _fetchInformation = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3() {
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                if (this.isRegistered()) {
                  _context3.next = 2;
                  break;
                }

                throw new NotRegisteredException();

              case 2:
                return _context3.abrupt("return", this.fetchJSON('GET', "/auth/register/".concat(this.oauthOptions.clientID), null, {
                  headers: {
                    Authorization: this.registrationAccessTokenToAuthHeader()
                  }
                }));

              case 3:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      return function fetchInformation() {
        return _fetchInformation.apply(this, arguments);
      };
    }()
    /**
     * Overwrites the client own information. This method will update both the local information and the remote information on the OAuth server.
     *
     * @throws {NotRegisteredException} When the client doesn't have it's registration information
     * @param   {object} information Set of information to update. Note that some fields such as `clientID` can't be updated.
     * @param   {boolean} resetSecret = false Optionnal, whether to reset the client secret or not
     * @returns {Promise} Resolves to a complete, updated list of client information
     */

  }, {
    key: "updateInformation",
    value: function () {
      var _updateInformation = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee4(information) {
        var resetSecret,
            mandatoryFields,
            data,
            result,
            _args4 = arguments;
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                resetSecret = _args4.length > 1 && _args4[1] !== undefined ? _args4[1] : false;

                if (this.isRegistered()) {
                  _context4.next = 3;
                  break;
                }

                throw new NotRegisteredException();

              case 3:
                mandatoryFields = {
                  clientID: this.oauthOptions.clientID,
                  clientName: this.oauthOptions.clientName,
                  redirectURI: this.oauthOptions.redirectURI,
                  softwareID: this.oauthOptions.softwareID
                };
                data = this.snakeCaseOAuthData((0, _objectSpread2.default)({}, mandatoryFields, information));
                if (resetSecret) data['client_secret'] = this.oauthOptions.clientSecret;
                _context4.next = 8;
                return this.fetchJSON('PUT', "/auth/register/".concat(this.oauthOptions.clientID), data, {
                  headers: {
                    Authorization: this.registrationAccessTokenToAuthHeader()
                  }
                });

              case 8:
                result = _context4.sent;
                this.setOAuthOptions((0, _objectSpread2.default)({}, data, result));
                return _context4.abrupt("return", this.oauthOptions);

              case 11:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      return function updateInformation(_x) {
        return _updateInformation.apply(this, arguments);
      };
    }()
    /**
     * Generates a random state code to be used during the OAuth process
     *
     * @returns {string}
     */

  }, {
    key: "generateStateCode",
    value: function generateStateCode() {
      var STATE_SIZE = 16;
      var hasCrypto = typeof window !== 'undefined' && typeof window.crypto !== 'undefined' && typeof window.crypto.getRandomValues === 'function';
      var buffer;

      if (hasCrypto) {
        buffer = new Uint8Array(STATE_SIZE);
        window.crypto.getRandomValues(buffer);
      } else {
        buffer = new Array(STATE_SIZE);

        for (var i = 0; i < buffer.length; i++) {
          buffer[i] = Math.floor(Math.random() * 255);
        }
      }

      return btoa(String.fromCharCode.apply(null, buffer)).replace(/=+$/, '').replace(/\//g, '_').replace(/\+/g, '-');
    }
    /**
     * Generates the URL that the user should be sent to in order to accept the app's permissions.
     *
     * @throws {NotRegisteredException} When the client doesn't have it's registration information
     * @param   {string} stateCode   A random code to be included in the URl for security. Can be generated with `client.generateStateCode()`
     * @param   {Array} scopes = [] An array of permission scopes for the token.
     * @returns {string} The URL
     */

  }, {
    key: "getAuthCodeURL",
    value: function getAuthCodeURL(stateCode) {
      var scopes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.scope;
      if (!this.isRegistered()) throw new NotRegisteredException();
      var query = {
        client_id: this.oauthOptions.clientID,
        redirect_uri: this.oauthOptions.redirectURI,
        state: stateCode,
        response_type: 'code',
        scope: scopes.join(' ')
      };
      return "".concat(this.uri, "/auth/authorize?").concat(this.dataToQueryString(query));
    }
  }, {
    key: "dataToQueryString",
    value: function dataToQueryString(data) {
      return Object.keys(data).map(function (param) {
        return "".concat(param, "=").concat(encodeURIComponent(data[param]));
      }).join('&');
    }
    /**
     * Retrieves the access code contained in the URL to which the user is redirected after accepting the app's permissions (the `redirectURI`).
     *
     * @throws {Error} The URL should contain the same state code as the one generated with `client.getAuthCodeURL()`. If not, it will throw an error
     * @param   {string} pageURL The redirected page URL, containing the state code and the access code
     * @param   {string} stateCode The state code that was contained in the original URL the user was sent to (see `client.getAuthCodeURL()`)
     * @returns {string} The access code
     */

  }, {
    key: "getAccessCodeFromURL",
    value: function getAccessCodeFromURL(pageURL, stateCode) {
      if (!stateCode) throw new Error('Missing state code');
      var params = new URL(pageURL).searchParams;
      var urlStateCode = params.get('state');
      var urlAccessCode = params.get('access_code');
      if (stateCode !== urlStateCode) throw new Error('Given state does not match url query state');
      return urlAccessCode;
    }
    /**
     * Exchanges an access code for an access token. This function does **not** update the client's token.
     *
     * @throws {NotRegisteredException} When the client doesn't have it's registration information
     * @param   {string} accessCode - The access code contained in the redirection URL — see `client.getAccessCodeFromURL()`
     * @param   {object} oauthOptionsArg — To use when OAuthClient is not yet registered (during login process)
     * @param   {string} uri — To use when OAuthClient is not yet registered (during login process)
     * @returns {Promise} A promise that resolves with an AccessToken object.
     */

  }, {
    key: "fetchAccessToken",
    value: function () {
      var _fetchAccessToken = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee5(accessCode, oauthOptionsArg, uri) {
        var oauthOptions, data, result;
        return _regenerator.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                if (!(!this.isRegistered() && !oauthOptionsArg)) {
                  _context5.next = 2;
                  break;
                }

                throw new NotRegisteredException();

              case 2:
                oauthOptions = oauthOptionsArg || this.oauthOptions;
                data = {
                  grant_type: 'authorization_code',
                  code: accessCode,
                  client_id: oauthOptions.clientID,
                  client_secret: oauthOptions.clientSecret
                };
                _context5.next = 6;
                return this.fetchJSON('POST', (uri || '') + '/auth/access_token', this.dataToQueryString(data), {
                  headers: {
                    'Content-Type': 'application/x-www-form-urlencoded'
                  }
                });

              case 6:
                result = _context5.sent;
                return _context5.abrupt("return", new _AccessToken.default(result));

              case 8:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      return function fetchAccessToken(_x2, _x3, _x4) {
        return _fetchAccessToken.apply(this, arguments);
      };
    }()
    /**
     * Retrieves a new access token by refreshing the currently used token.
     *
     * @throws {NotRegisteredException} When the client doesn't have it's registration information
     * @throws {Error} The client should already have an access token to use this function
     * @returns {Promise} A promise that resolves with a new AccessToken object
     */

  }, {
    key: "refreshToken",
    value: function () {
      var _refreshToken = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee6() {
        var data, result, newToken;
        return _regenerator.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                if (this.isRegistered()) {
                  _context6.next = 2;
                  break;
                }

                throw new NotRegisteredException();

              case 2:
                if (this.token) {
                  _context6.next = 4;
                  break;
                }

                throw new Error('No token to refresh');

              case 4:
                data = {
                  grant_type: 'refresh_token',
                  refresh_token: this.token.refreshToken,
                  client_id: this.oauthOptions.clientID,
                  client_secret: this.oauthOptions.clientSecret
                };
                _context6.next = 7;
                return (0, _get2.default)((0, _getPrototypeOf2.default)(OAuthClient.prototype), "fetchJSON", this).call(this, 'POST', '/auth/access_token', this.dataToQueryString(data), {
                  headers: {
                    'Content-Type': 'application/x-www-form-urlencoded'
                  }
                });

              case 7:
                result = _context6.sent;
                newToken = new _AccessToken.default((0, _objectSpread2.default)({
                  refresh_token: this.token.refreshToken
                }, result));

                if (this.onTokenRefresh && typeof this.onTokenRefresh === 'function') {
                  this.onTokenRefresh(newToken);
                }

                return _context6.abrupt("return", newToken);

              case 11:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      return function refreshToken() {
        return _refreshToken.apply(this, arguments);
      };
    }()
  }, {
    key: "exchangeOAuthSecret",
    value: function exchangeOAuthSecret(uri, secret) {
      return this.fetchJSON('POST', uri + '/auth/secret_exchange', {
        secret: secret
      });
    }
    /**
     * Updates the client's stored token
     *
     * @param {string} token = null The new token to use — can be a string, a json object or an AccessToken instance.
     */

  }, {
    key: "setToken",
    value: function setToken(token) {
      if (token) {
        this.token = token instanceof _AccessToken.default ? token : new _AccessToken.default(token);
      } else {
        this.token = null;
      }
    }
  }, {
    key: "setCredentials",
    value: function setCredentials(token) {
      (0, _logDeprecate.default)('setCredentials is deprecated, please replace by setToken');
      return this.setToken(token);
    }
    /**
     * Updates the OAuth informations
     *
     * @param {object} options Map of OAuth options
     */

  }, {
    key: "setOAuthOptions",
    value: function setOAuthOptions(options) {
      this.oauthOptions = this.camelCaseOAuthData(options);
    }
  }, {
    key: "resetClientId",
    value: function resetClientId() {
      this.oauthOptions.clientID = '';
    }
    /**
     * Reset the current OAuth client
     */

  }, {
    key: "resetClient",
    value: function resetClient() {
      this.resetClientId();
      this.setUri(null);
      this.setToken(null);
    }
    /**
     * Turns the client's registration access token into a header suitable for HTTP requests. Used in some queries to manipulate the client on the server side.
     *
     * @returns {string}
     * @private
     */

  }, {
    key: "registrationAccessTokenToAuthHeader",
    value: function registrationAccessTokenToAuthHeader() {
      if (!this.oauthOptions.registrationAccessToken) {
        throw new Error('No registration access token');
      }

      return 'Bearer ' + this.oauthOptions.registrationAccessToken;
    }
  }]);
  return OAuthClient;
}(_CozyStackClient2.default);

var NotRegisteredException =
/*#__PURE__*/
function (_Error) {
  (0, _inherits2.default)(NotRegisteredException, _Error);

  function NotRegisteredException() {
    var _this2;

    var message = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Client not registered or missing OAuth information';
    (0, _classCallCheck2.default)(this, NotRegisteredException);
    _this2 = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(NotRegisteredException).call(this, message));
    _this2.message = message;
    _this2.name = 'NotRegisteredException';
    return _this2;
  }

  return NotRegisteredException;
}((0, _wrapNativeSuper2.default)(Error));

var _default = OAuthClient;
exports.default = _default;
