// cozy-stack-client/dist/PermissionCollection.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.getPermissionsFor = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _DocumentCollection2 = _interopRequireWildcard(require("./DocumentCollection"));

var _FileCollection = require("./FileCollection");

var _utils = require("./utils");

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2.default)(["/permissions/doctype/", "/shared-by-link"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2.default)(["/permissions/", ""]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2.default)(["/permissions/", ""]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var normalizePermission = function normalizePermission(perm) {
  return (0, _DocumentCollection2.normalizeDoc)(perm, 'io.cozy.permissions');
};
/**
 * Implements `DocumentCollection` API along with specific methods for `io.cozy.permissions`.
 */


var PermissionCollection =
/*#__PURE__*/
function (_DocumentCollection) {
  (0, _inherits2.default)(PermissionCollection, _DocumentCollection);

  function PermissionCollection() {
    (0, _classCallCheck2.default)(this, PermissionCollection);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(PermissionCollection).apply(this, arguments));
  }

  (0, _createClass2.default)(PermissionCollection, [{
    key: "get",
    value: function () {
      var _get = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(id) {
        var resp;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.stackClient.fetchJSON('GET', (0, _utils.uri)(_templateObject(), id));

              case 2:
                resp = _context.sent;
                return _context.abrupt("return", {
                  data: normalizePermission(resp.data)
                });

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function get(_x) {
        return _get.apply(this, arguments);
      };
    }()
    /**
     * Create a new set of permissions
     * It can also associates one or more codes to it, via the codes parameter
     *
     * @param {object} permission
     * @param {string} permission.codes A comma separed list of values (defaulted to code)
     * @param {string} permission.ttl Make the codes expire after a delay (bigduration format)
     *
     * bigduration format: https://github.com/justincampbell/bigduration/blob/master/README.md
     * @see https://docs.cozy.io/en/cozy-stack/permissions/#post-permissions
     *
     */

  }, {
    key: "create",
    value: function () {
      var _create = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(_ref) {
        var _id, _type, _ref$codes, codes, ttl, attributes, searchParams, resp;

        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _id = _ref._id, _type = _ref._type, _ref$codes = _ref.codes, codes = _ref$codes === void 0 ? 'code' : _ref$codes, ttl = _ref.ttl, attributes = (0, _objectWithoutProperties2.default)(_ref, ["_id", "_type", "codes", "ttl"]);
                searchParams = new URLSearchParams();
                searchParams.append('codes', codes);
                if (ttl) searchParams.append('ttl', ttl);
                _context2.next = 6;
                return this.stackClient.fetchJSON('POST', "/permissions?".concat(searchParams), {
                  data: {
                    type: 'io.cozy.permissions',
                    attributes: attributes
                  }
                });

              case 6:
                resp = _context2.sent;
                return _context2.abrupt("return", {
                  data: normalizePermission(resp.data)
                });

              case 8:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function create(_x2) {
        return _create.apply(this, arguments);
      };
    }()
    /**
     * Adds a permission to the given document. Document type must be
     * `io.cozy.apps`, `io.cozy.konnectors` or `io.cozy.permissions`
     *
     * @param  {object}  document - Document which receives the permission
     * @param  {object}  permission - Describes the permission
     * @returns {Promise}
     *
     * @example
     * ```
     * const permissions = await client
     *   .collection('io.cozy.permissions')
     *   .add(konnector, {
     *     folder: {
     *       type: 'io.cozy.files',
     *       verbs: ['GET', 'PUT'],
     *       values: [`io.cozy.files.bc57b60eb2954537b0dcdc6ebd8e9d23`]
     *     }
     *  })
     * ```
     */

  }, {
    key: "add",
    value: function () {
      var _add = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3(document, permission) {
        var endpoint, resp;
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.t0 = document._type;
                _context3.next = _context3.t0 === 'io.cozy.apps' ? 3 : _context3.t0 === 'io.cozy.konnectors' ? 5 : _context3.t0 === 'io.cozy.permissions' ? 7 : 9;
                break;

              case 3:
                endpoint = "/permissions/apps/".concat(document.slug);
                return _context3.abrupt("break", 10);

              case 5:
                endpoint = "/permissions/konnectors/".concat(document.slug);
                return _context3.abrupt("break", 10);

              case 7:
                endpoint = "/permissions/".concat(document._id);
                return _context3.abrupt("break", 10);

              case 9:
                throw new Error('Permissions can only be added on existing permissions, apps and konnectors.');

              case 10:
                _context3.next = 12;
                return this.stackClient.fetchJSON('PATCH', endpoint, {
                  data: {
                    type: 'io.cozy.permissions',
                    attributes: {
                      permissions: permission
                    }
                  }
                });

              case 12:
                resp = _context3.sent;
                return _context3.abrupt("return", {
                  data: normalizePermission(resp.data)
                });

              case 14:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      return function add(_x3, _x4) {
        return _add.apply(this, arguments);
      };
    }()
  }, {
    key: "destroy",
    value: function destroy(permission) {
      return this.stackClient.fetchJSON('DELETE', (0, _utils.uri)(_templateObject2(), permission.id));
    }
  }, {
    key: "findLinksByDoctype",
    value: function () {
      var _findLinksByDoctype = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee4(doctype) {
        var resp;
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.stackClient.fetchJSON('GET', (0, _utils.uri)(_templateObject3(), doctype));

              case 2:
                resp = _context4.sent;
                return _context4.abrupt("return", (0, _objectSpread2.default)({}, resp, {
                  data: resp.data.map(normalizePermission)
                }));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      return function findLinksByDoctype(_x5) {
        return _findLinksByDoctype.apply(this, arguments);
      };
    }()
  }, {
    key: "findApps",
    value: function () {
      var _findApps = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee5() {
        var resp;
        return _regenerator.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this.stackClient.fetchJSON('GET', '/apps/');

              case 2:
                resp = _context5.sent;
                return _context5.abrupt("return", (0, _objectSpread2.default)({}, resp, {
                  data: resp.data.map(function (a) {
                    return (0, _objectSpread2.default)({
                      _id: a.id
                    }, a);
                  })
                }));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      return function findApps() {
        return _findApps.apply(this, arguments);
      };
    }()
    /**
     * Create a share link
     *
     * @param {{_id, _type}} document - cozy document
     * @param {object} options - options
     * @param {string[]} options.verbs - explicit permissions to use
     */

  }, {
    key: "createSharingLink",
    value: function () {
      var _createSharingLink = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee6(document) {
        var options,
            verbs,
            resp,
            _args6 = arguments;
        return _regenerator.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                options = _args6.length > 1 && _args6[1] !== undefined ? _args6[1] : {};
                verbs = options.verbs;
                _context6.next = 4;
                return this.stackClient.fetchJSON('POST', "/permissions?codes=email", {
                  data: {
                    type: 'io.cozy.permissions',
                    attributes: {
                      permissions: getPermissionsFor(document, true, verbs ? {
                        verbs: verbs
                      } : {})
                    }
                  }
                });

              case 4:
                resp = _context6.sent;
                return _context6.abrupt("return", {
                  data: normalizePermission(resp.data)
                });

              case 6:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      return function createSharingLink(_x6) {
        return _createSharingLink.apply(this, arguments);
      };
    }()
    /**
     * Follow the next link to fetch the next permissions
     *
     * @param {object} permissions JSON-API based permissions document
     */

  }, {
    key: "fetchPermissionsByLink",
    value: function () {
      var _fetchPermissionsByLink = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee7(permissions) {
        return _regenerator.default.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                if (!(permissions.links && permissions.links.next)) {
                  _context7.next = 4;
                  break;
                }

                _context7.next = 3;
                return this.stackClient.fetchJSON('GET', permissions.links.next);

              case 3:
                return _context7.abrupt("return", _context7.sent);

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      return function fetchPermissionsByLink(_x7) {
        return _fetchPermissionsByLink.apply(this, arguments);
      };
    }()
    /**
     *
     * @param {object} document Cozy doc
     * @returns {object} with all the permissions
     */

  }, {
    key: "fetchAllLinks",
    value: function () {
      var _fetchAllLinks = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee8(document) {
        var allLinks, resp, _allLinks$data;

        return _regenerator.default.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this.findLinksByDoctype(document._type);

              case 2:
                allLinks = _context8.sent;
                resp = allLinks;

              case 4:
                if (!(resp.links && resp.links.next)) {
                  _context8.next = 11;
                  break;
                }

                _context8.next = 7;
                return this.fetchPermissionsByLink(resp);

              case 7:
                resp = _context8.sent;

                (_allLinks$data = allLinks.data).push.apply(_allLinks$data, (0, _toConsumableArray2.default)(resp.data.map(normalizePermission)));

                _context8.next = 4;
                break;

              case 11:
                return _context8.abrupt("return", allLinks);

              case 12:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      return function fetchAllLinks(_x8) {
        return _fetchAllLinks.apply(this, arguments);
      };
    }()
    /**
     * Destroy a sharing link and the related permissions
     *
     * @param {object} document
     */

  }, {
    key: "revokeSharingLink",
    value: function () {
      var _revokeSharingLink = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee9(document) {
        var allLinks, links, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, perm;

        return _regenerator.default.wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this.fetchAllLinks(document);

              case 2:
                allLinks = _context9.sent;
                links = allLinks.data.filter(function (perm) {
                  return isPermissionRelatedTo(perm, document);
                });
                _iteratorNormalCompletion = true;
                _didIteratorError = false;
                _iteratorError = undefined;
                _context9.prev = 7;
                _iterator = links[Symbol.iterator]();

              case 9:
                if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
                  _context9.next = 16;
                  break;
                }

                perm = _step.value;
                _context9.next = 13;
                return this.destroy(perm);

              case 13:
                _iteratorNormalCompletion = true;
                _context9.next = 9;
                break;

              case 16:
                _context9.next = 22;
                break;

              case 18:
                _context9.prev = 18;
                _context9.t0 = _context9["catch"](7);
                _didIteratorError = true;
                _iteratorError = _context9.t0;

              case 22:
                _context9.prev = 22;
                _context9.prev = 23;

                if (!_iteratorNormalCompletion && _iterator.return != null) {
                  _iterator.return();
                }

              case 25:
                _context9.prev = 25;

                if (!_didIteratorError) {
                  _context9.next = 28;
                  break;
                }

                throw _iteratorError;

              case 28:
                return _context9.finish(25);

              case 29:
                return _context9.finish(22);

              case 30:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this, [[7, 18, 22, 30], [23,, 25, 29]]);
      }));

      return function revokeSharingLink(_x9) {
        return _revokeSharingLink.apply(this, arguments);
      };
    }()
    /**
     * async getOwnPermissions - Gets the permission for the current token
     *
     * @returns {object}
     */

  }, {
    key: "getOwnPermissions",
    value: function () {
      var _getOwnPermissions = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee10() {
        var resp;
        return _regenerator.default.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this.stackClient.fetchJSON('GET', '/permissions/self');

              case 2:
                resp = _context10.sent;
                return _context10.abrupt("return", {
                  data: normalizePermission(resp.data)
                });

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      return function getOwnPermissions() {
        return _getOwnPermissions.apply(this, arguments);
      };
    }()
  }]);
  return PermissionCollection;
}(_DocumentCollection2.default);
/**
 * Build a permission set
 *
 * @param {{_id, _type}} document - cozy document
 * @param {boolean} publicLink - are the permissions for a public link ?
 * @param {object} options - options
 * @param {string[]} options.verbs - explicit permissions to use
 * @returns {object} permissions object that can be sent through /permissions/*
 */


var getPermissionsFor = function getPermissionsFor(document) {
  var publicLink = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var _id = document._id,
      _type = document._type;
  var verbs = options.verbs ? options.verbs : publicLink ? ['GET'] : ['ALL']; // TODO: this works for albums, but it needs to be generalized and integrated
  // with cozy-client ; some sort of doctype "schema" will be needed here

  return (0, _FileCollection.isFile)(document) ? {
    files: {
      type: 'io.cozy.files',
      verbs: verbs,
      values: [_id]
    }
  } : {
    collection: {
      type: _type,
      verbs: verbs,
      values: [_id]
    },
    files: {
      type: 'io.cozy.files',
      verbs: verbs,
      values: ["".concat(_type, "/").concat(_id)],
      selector: 'referenced_by'
    }
  };
};

exports.getPermissionsFor = getPermissionsFor;
PermissionCollection.normalizeDoctype = _DocumentCollection2.default.normalizeDoctypeJsonApi;

var isPermissionRelatedTo = function isPermissionRelatedTo(perm, document) {
  var _id = document._id;
  return (0, _FileCollection.isFile)(document) ? perm.attributes.permissions.files.values.indexOf(_id) !== -1 : perm.attributes.permissions.collection.values.indexOf(_id) !== -1;
};

var _default = PermissionCollection;
exports.default = _default;
