// cozy-stack-client/dist/ShortcutsCollection.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SHORTCUTS_DOCTYPE = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _DocumentCollection2 = _interopRequireDefault(require("./DocumentCollection"));

var _utils = require("./utils");

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2.default)(["/shortcuts/", ""]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2.default)(["/shortcuts"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var SHORTCUTS_DOCTYPE = 'io.cozy.files.shortcuts';
exports.SHORTCUTS_DOCTYPE = SHORTCUTS_DOCTYPE;

var ShortcutsCollection =
/*#__PURE__*/
function (_DocumentCollection) {
  (0, _inherits2.default)(ShortcutsCollection, _DocumentCollection);

  function ShortcutsCollection(stackClient) {
    (0, _classCallCheck2.default)(this, ShortcutsCollection);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ShortcutsCollection).call(this, SHORTCUTS_DOCTYPE, stackClient));
  }
  /**
   * Create a shortcut
   *
   * @param {object} attributes shortcut's attributes
   * @param {string} attributes.name Filename
   * @param {string} attributes.url Shortcut's URL
   * @param {string} attributes.dir_id dir_id where to create the shortcut
   */


  (0, _createClass2.default)(ShortcutsCollection, [{
    key: "create",
    value: function () {
      var _create = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(attributes) {
        var path, resp;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!attributes.type) {
                  attributes.type = SHORTCUTS_DOCTYPE;
                }

                if (!(!attributes.name || !attributes.url || !attributes.dir_id)) {
                  _context.next = 3;
                  break;
                }

                throw new Error('you need at least a name, an url and a dir_id attributes to create a shortcut');

              case 3:
                path = (0, _utils.uri)(_templateObject());
                _context.next = 6;
                return this.stackClient.fetchJSON('POST', path, {
                  data: {
                    attributes: attributes,
                    type: 'io.cozy.files.shortcuts'
                  }
                });

              case 6:
                resp = _context.sent;
                return _context.abrupt("return", {
                  data: _DocumentCollection2.default.normalizeDoctypeJsonApi(SHORTCUTS_DOCTYPE)(resp.data, resp)
                });

              case 8:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function create(_x) {
        return _create.apply(this, arguments);
      };
    }()
  }, {
    key: "get",
    value: function () {
      var _get = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(id) {
        var path, resp;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                path = (0, _utils.uri)(_templateObject2(), id);
                _context2.next = 3;
                return this.stackClient.fetchJSON('GET', path);

              case 3:
                resp = _context2.sent;
                return _context2.abrupt("return", {
                  data: _DocumentCollection2.default.normalizeDoctypeJsonApi(SHORTCUTS_DOCTYPE)(resp.data, resp)
                });

              case 5:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function get(_x2) {
        return _get.apply(this, arguments);
      };
    }()
  }]);
  return ShortcutsCollection;
}(_DocumentCollection2.default);

ShortcutsCollection.normalizeDoctype = _DocumentCollection2.default.normalizeDoctypeJsonApi;
var _default = ShortcutsCollection;
exports.default = _default;
