// cozy-client/dist/__tests__/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createTestAssets = exports.queryResultFromData = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _redux = require("redux");

var _CozyLink = _interopRequireDefault(require("../CozyLink"));

var _CozyClient = _interopRequireDefault(require("../CozyClient"));

// TODO
// Provide a better means to test with the client
// See also ./mocks.js
var queryResultFromData = function queryResultFromData(data) {
  var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return (0, _objectSpread2.default)({
    data: data,
    meta: {
      count: data.length
    },
    skip: 0,
    next: false
  }, opts);
};

exports.queryResultFromData = queryResultFromData;

var createTestAssets = function createTestAssets() {
  // TODO this requestHandler should be improved upon
  var requestHandler = function requestHandler() {
    return {
      data: []
    };
  };

  var link = new _CozyLink.default(requestHandler);
  var client = new _CozyClient.default({
    links: [link]
  });
  var store = (0, _redux.createStore)((0, _redux.combineReducers)({
    cozy: client.reducer()
  }));
  client.setStore(store);
  return {
    client: client,
    store: store,
    link: link
  };
};

exports.createTestAssets = createTestAssets;
