// cozy-client/dist/hooks/useQuery.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.useQueries = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _react = require("react");

var _reactRedux = require("react-redux");

var _get = _interopRequireDefault(require("lodash/get"));

var _useClient = _interopRequireDefault(require("./useClient"));

var _logger = _interopRequireDefault(require("../logger"));

var resolveToValue = function resolveToValue(fnOrValue) {
  return typeof fnOrValue === 'function' ? fnOrValue() : fnOrValue;
};

var generateFetchMoreQueryDefinition = function generateFetchMoreQueryDefinition(queryResult) {
  return queryResult.bookmark ? queryResult.definition.offsetBookmark(queryResult.bookmark) : queryResult.definition.offset(queryResult.data.length);
};
/**
 * Fetches a queryDefinition and returns the queryState
 *
 * @param  {object} queryDefinition - Definition created with Q()
 *
 * @param  {object} options - Options
 * @param  {object} options.as - Name for the query [required]
 * @param  {object} options.fetchPolicy - Fetch policy
 * @param  {object} options.singleDocData - If true, the "data" returned will be
 * a single doc instead of an array for single doc queries. Defaults to false for backward
 * compatibility but will be set to true in the future.
 *
 * @returns {object}
 */


var useQuery = function useQuery(queryDefinition, options) {
  if (!_reactRedux.useSelector) {
    throw new Error('You must use react-redux > 7.1.0 to use useQuery (uses useSelector) under the hood');
  }

  if (!queryDefinition) {
    _logger.default.warn('Bad query', queryDefinition);

    throw new Error('Bad query');
  }

  var definition = resolveToValue(queryDefinition);
  var as = options.as;

  if (!as) {
    throw new Error('You must specify options.as when using useQuery');
  }

  var client = (0, _useClient.default)();
  var queryState = (0, _reactRedux.useSelector)(function () {
    if (options.singleDocData === undefined) {
      _logger.default.warn('useQuery options.singleDocData will pass to true in a next version of cozy-client, please add it now to prevent any problem in the future.');
    }

    return client.getQueryFromState(as, {
      hydrated: (0, _get.default)(options, 'hydrated', true),
      singleDocData: (0, _get.default)(options, 'singleDocData', false)
    });
  });
  (0, _react.useEffect)(function () {
    client.query(definition, options);
  }, // eslint-disable-next-line react-hooks/exhaustive-deps
  [as]);
  var fetchMore = (0, _react.useCallback)(function () {
    var queryState = client.getQueryFromState(as);
    return client.query(generateFetchMoreQueryDefinition(queryState), {
      as: as
    });
  }, [as, client]);
  return (0, _objectSpread2.default)({}, queryState, {
    fetchMore: fetchMore
  });
};

var useQueries = function useQueries(querySpecs) {
  var res = {};

  for (var _i = 0, _Object$entries = Object.entries(querySpecs); _i < _Object$entries.length; _i++) {
    var _Object$entries$_i = (0, _slicedToArray2.default)(_Object$entries[_i], 2),
        queryAttrName = _Object$entries$_i[0],
        queryOpts = _Object$entries$_i[1];

    // eslint-disable-next-line
    res[queryAttrName] = useQuery(queryOpts.query, queryOpts);
  }

  return res;
};

exports.useQueries = useQueries;
var _default = useQuery;
exports.default = _default;
