// cozy-client/dist/associations/HasMany.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.updateRelationship = exports.updateHasManyItem = exports.setHasManyItem = exports.getHasManyItem = void 0;

var _objectSpread5 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _get = _interopRequireDefault(require("lodash/get"));

var _merge = _interopRequireDefault(require("lodash/merge"));

var _Association2 = _interopRequireDefault(require("./Association"));

var _dsl = require("../queries/dsl");

var _store = require("../store");

var empty = function empty() {
  return {
    data: [],
    next: true,
    meta: {
      count: 0
    }
  };
};

var updateArray = function updateArray(array, indexArg, el) {
  var index = indexArg === -1 ? array.length : indexArg;
  return [].concat((0, _toConsumableArray2.default)(array.slice(0, index)), [el], (0, _toConsumableArray2.default)(array.slice(index + 1)));
};
/**
 * Related documents are stored in the relationships attribute of the object,
 * following the JSON API spec.
 *
 * Responsible for
 *
 * - Creating relationships
 * - Removing relationships
 *
 * @description
 *
 * ```
 * const schema = {
 *   todos: {
 *      doctype: 'io.cozy.todos',
 *      relationships: {
 *        tasks: {
 *          doctype: 'io.cozy.tasks',
 *          type: 'has-many'
 *        }
 *      }
 *    }
 * }
 *
 * const todo = {
 *   label: "Protect people's privacy",
 *   relationships: {
 *     tasks: {
 *       data: [
 *         {_id: 1, _type: 'io.cozy.tasks'},
 *         {_id: 2, _type: 'io.cozy.tasks'}
 *       ]
 *     }
 *   }
 * }
 * ```
 */


var HasMany =
/*#__PURE__*/
function (_Association) {
  (0, _inherits2.default)(HasMany, _Association);

  function HasMany() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, HasMany);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(HasMany)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "updateRelationshipData", function (getUpdatedRelationshipData) {
      return function (dispatch, getState) {
        var previousRelationship = (0, _store.getDocumentFromState)(getState(), _this.target._type, _this.target._id);
        dispatch((0, _store.receiveQueryResult)(null, {
          data: (0, _objectSpread5.default)({}, previousRelationship, {
            relationships: (0, _objectSpread5.default)({}, previousRelationship.relationships, (0, _defineProperty2.default)({}, _this.name, getUpdatedRelationshipData(previousRelationship.relationships[_this.name])))
          })
        }));
      };
    });
    return _this;
  }

  (0, _createClass2.default)(HasMany, [{
    key: "fetchMore",
    value: function fetchMore() {
      throw 'Not implemented';
    }
  }, {
    key: "exists",
    value: function exists(document) {
      return this.existsById(document._id);
    }
  }, {
    key: "containsById",
    value: function containsById(id) {
      return this.getRelationship().data.find(function (_ref) {
        var _id = _ref._id;
        return id === _id;
      }) !== undefined;
    }
  }, {
    key: "existsById",
    value: function existsById(id) {
      return this.containsById(id) && Boolean(this.get(this.doctype, id));
    }
    /**
     * Add a referenced document by id. You need to call save()
     * in order to synchronize your document with the store.
     *
     * @todo We shouldn't create the array of relationship manually since
     * it'll not be present in the store as well.
     * We certainly should use something like `updateRelationship`
     *
     */

  }, {
    key: "addById",
    value: function addById(idsArg) {
      var _this2 = this,
          _this$target$relation;

      if (!this.target.relationships) this.target.relationships = {};

      if (!this.target.relationships[this.name]) {
        this.target.relationships[this.name] = {
          data: []
        };
      }

      var ids = Array.isArray(idsArg) ? idsArg : [idsArg];
      var newRelations = ids.filter(function (id) {
        return !_this2.existsById(id);
      }).map(function (id) {
        return {
          _id: id,
          _type: _this2.doctype
        };
      });

      (_this$target$relation = this.target.relationships[this.name].data).push.apply(_this$target$relation, (0, _toConsumableArray2.default)(newRelations));

      this.updateMetaCount();
      return this.save(this.target);
    }
  }, {
    key: "removeById",
    value: function removeById(idsArg) {
      var ids = Array.isArray(idsArg) ? idsArg : [idsArg];
      this.target.relationships[this.name].data = this.target.relationships[this.name].data.filter(function (_ref2) {
        var _id = _ref2._id;
        return !ids.includes(_id);
      });
      this.updateMetaCount();
      return this.save(this.target);
    }
  }, {
    key: "updateMetaCount",
    value: function updateMetaCount() {
      if ((0, _get.default)(this.target.relationships[this.name], 'meta.count') !== undefined) {
        this.target.relationships[this.name].meta = (0, _objectSpread5.default)({}, this.target.relationships[this.name].meta, {
          count: this.target.relationships[this.name].data.length
        });
      }
    }
  }, {
    key: "getRelationship",
    value: function getRelationship() {
      var rawData = this.target[this.name];
      var relationship = (0, _get.default)(this.target, "relationships.".concat(this.name));

      if (!relationship) {
        if (rawData && rawData.length) {
          console.warn("You're trying to access data on a relationship that appear to not be loaded yet. You may want to use 'include()' on your query");
        }

        return empty();
      }

      return relationship;
    }
  }, {
    key: "updateTargetRelationship",
    value: function updateTargetRelationship(store, updateFn) {
      // TODO See if updateTargetRelationship is still used, removing it would enable us
      // to remove store.readDocument and store.writeDocument and the StoreProxy
      var prevTarget = store.readDocument(this.target._type, this.target._id);
      store.writeDocument(this.updateRelationship(prevTarget, updateFn));
    }
  }, {
    key: "updateRelationship",
    value: function updateRelationship(target, updateFn) {
      return HasMany.updateRelationship(target, this.name, updateFn);
    }
  }, {
    key: "dehydrate",
    value: function dehydrate(doc) {
      return (0, _objectSpread5.default)({}, doc, {
        relationships: (0, _objectSpread5.default)({}, doc.relationships, (0, _defineProperty2.default)({}, this.name, {
          data: this.raw
        }))
      });
    }
  }, {
    key: "raw",
    get: function get() {
      return this.getRelationship().data;
    }
  }, {
    key: "data",
    get: function get() {
      var _this3 = this;

      return this.getRelationship().data.map(function (_ref3) {
        var _id = _ref3._id,
            _type = _ref3._type;
        return _this3.get(_type, _id);
      }).filter(Boolean);
    }
  }, {
    key: "hasMore",
    get: function get() {
      return this.getRelationship().next;
    }
    /**
     * Returns the total number of documents in the relationship.
     * Does not handle documents absent from the store. If you want
     * to do that, you can use .data.length.
     *
     * @returns {number} - Total number of documents in the relationships
     */

  }, {
    key: "count",
    get: function get() {
      var relationship = this.getRelationship();
      return relationship.meta ? relationship.meta.count : relationship.data.length;
    }
  }], [{
    key: "query",
    value: function query(document, client, assoc) {
      var relationships = (0, _get.default)(document, "relationships.".concat(assoc.name, ".data"), []);
      var ids = relationships.map(function (assoc) {
        return assoc._id;
      });
      return new _dsl.QueryDefinition({
        doctype: assoc.doctype,
        ids: ids
      });
    }
  }]);
  return HasMany;
}(_Association2.default);
/**
 * Gets a relationship item with the relationship name and id
 *
 * @param {object} doc - Document to be updated
 * @param {string} relName - Name of the relationship
 * @param {string} relItemId - Id of the relationship item
 */


var getHasManyItem = HasMany.getHasManyItem = function (doc, relName, relItemId) {
  var relData = (0, _get.default)(doc, "relationships.".concat(relName, ".data"), []);
  return relData.find(function (rel) {
    return rel._id == relItemId;
  });
};
/**
 * Sets a relationship item with the relationship name and id
 *
 * @param {object} doc - Document to be updated
 * @param {string} relName - Name of the relationship
 * @param {string} relItemId - Id of the relationship item
 * @param {object} relItemAttrs - Attributes to be set (at least _id and _type)
 */


exports.getHasManyItem = getHasManyItem;

var setHasManyItem = HasMany.setHasManyItem = function (doc, relName, relItemId, relItemAttrs) {
  var relData = (0, _get.default)(doc, "relationships.".concat(relName, ".data"), []);
  var relIndex = relData.findIndex(function (rel) {
    return rel._id === relItemId;
  });
  var updatedRelItem = (0, _merge.default)({}, relData[relIndex], relItemAttrs);
  var updatedRelData = updateArray(relData, relIndex, updatedRelItem);
  var updatedDocument = HasMany.updateRelationship(doc, relName, function (relationship) {
    return (0, _merge.default)({}, relationship, {
      data: updatedRelData
    });
  });
  return updatedDocument;
};
/**
 * Updates a relationship item with the relationship name and id
 *
 * @param {object} doc - Document to be updated
 * @param {string} relName - Name of the relationship
 * @param {string} relItemId - Id of the relationship item
 * @param {Function} updater - receives the current relationship item and should
 * return an updated version. Merge should be used in the updater
 * if previous relationship item fields are to be kept.
 */


exports.setHasManyItem = setHasManyItem;

var updateHasManyItem = HasMany.updateHasManyItem = function (doc, relName, relItemId, updater) {
  var relItem = HasMany.getHasManyItem(doc, relName, relItemId);
  var updatedRelItem = updater(relItem);
  return HasMany.setHasManyItem(doc, relName, relItemId, updatedRelItem);
};

exports.updateHasManyItem = updateHasManyItem;

var updateRelationship = HasMany.updateRelationship = function (doc, relName, updateFn) {
  return (0, _objectSpread5.default)({}, doc, {
    relationships: (0, _objectSpread5.default)({}, doc.relationships, (0, _defineProperty2.default)({}, relName, (0, _objectSpread5.default)({}, doc.relationships[relName], updateFn(doc.relationships[name]))))
  });
};

exports.updateRelationship = updateRelationship;
var _default = HasMany;
exports.default = _default;
