// cozy-client/dist/associations/HasManyInPlace.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectSpread3 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _Association2 = _interopRequireDefault(require("./Association"));

var _dsl = _interopRequireDefault(require("../queries/dsl"));

/**
 *
 * Used when related documents are stored directly under the attribute with
 * only the ids.
 *
 * @description
 *
 * An example document representing a TODO. See as the related
 * tasks are represented via ids.
 *
 * ```js
 * const todo = {
 *   label: "Protect people's privacy",
 *   tasks: [1, 2]
 * }
 * ```
 *
 * Here is the `Schema` that would represent this kind of document.
 * Components receiving todos via `Query`s would have an instance of `HasManyInPlace`
 * as their `tasks` attribute.
 *
 * ```js
 * const schema = {
 *   todos: {
 *      doctype: 'io.cozy.todos',
 *      relationships: {
 *        tasks: {
 *          doctype: 'io.cozy.tasks',
 *          type: 'has-many-in-place'
 *        }
 *      }
 *    }
 * }
 *
 * const todo = {
 *   label: "Get rich",
 *   tasks: [1, 2]
 * }
 * ```
 *
 */
var HasManyInPlace =
/*#__PURE__*/
function (_Association) {
  (0, _inherits2.default)(HasManyInPlace, _Association);

  function HasManyInPlace() {
    (0, _classCallCheck2.default)(this, HasManyInPlace);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(HasManyInPlace).apply(this, arguments));
  }

  (0, _createClass2.default)(HasManyInPlace, [{
    key: "addById",
    value: function addById(id) {
      var rel = this.getRelationship();
      rel.push(id);
    }
  }, {
    key: "removeById",
    value: function removeById(id) {
      var rel = this.getRelationship();
      var index = rel.indexOf(id);

      if (index !== -1) {
        rel.splice(index, 1);
      }
    }
  }, {
    key: "existsById",
    value: function existsById(id) {
      var rel = this.getRelationship();
      return rel.indexOf(id) !== -1;
    }
  }, {
    key: "getRelationship",
    value: function getRelationship() {
      this.target[this.name] = this.target[this.name] || [];
      return this.target[this.name];
    }
  }, {
    key: "dehydrate",
    value: function dehydrate(doc) {
      return (0, _objectSpread3.default)({}, doc, (0, _defineProperty2.default)({}, this.name, this.raw || []));
    }
  }, {
    key: "raw",
    get: function get() {
      return this.target[this.name];
    }
  }, {
    key: "data",
    get: function get() {
      var _this = this;

      var doctype = this.doctype;
      return (this.raw || []).map(function (_id) {
        return _this.get(doctype, _id);
      });
    }
  }], [{
    key: "query",
    value: function query() {
      if (this.raw && this.raw.length > 0) {
        return (0, _dsl.default)({
          doctype: this.doctype,
          ids: this.raw
        });
      } else {
        return null;
      }
    }
  }]);
  return HasManyInPlace;
}(_Association2.default);

var _default = HasManyInPlace;
exports.default = _default;
