// cozy-client/dist/store/mutations.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.receiveMutationError = exports.receiveMutationResult = exports.initMutation = exports.isReceivingMutationResult = exports.isMutationAction = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var INIT_MUTATION = 'INIT_MUTATION';
var RECEIVE_MUTATION_RESULT = 'RECEIVE_MUTATION_RESULT';
var RECEIVE_MUTATION_ERROR = 'RECEIVE_MUTATION_ERROR';

var isMutationAction = function isMutationAction(action) {
  return [INIT_MUTATION, RECEIVE_MUTATION_RESULT, RECEIVE_MUTATION_ERROR].indexOf(action.type) !== -1;
};

exports.isMutationAction = isMutationAction;

var isReceivingMutationResult = function isReceivingMutationResult(action) {
  return action.type === RECEIVE_MUTATION_RESULT;
}; // actions


exports.isReceivingMutationResult = isReceivingMutationResult;

var initMutation = function initMutation(mutationId, definition) {
  return {
    type: INIT_MUTATION,
    mutationId: mutationId,
    definition: definition
  };
};

exports.initMutation = initMutation;

var receiveMutationResult = function receiveMutationResult(mutationId, response) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var definition = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  return (0, _objectSpread2.default)({
    type: RECEIVE_MUTATION_RESULT,
    mutationId: mutationId,
    response: response
  }, options, {
    definition: definition
  });
};

exports.receiveMutationResult = receiveMutationResult;

var receiveMutationError = function receiveMutationError(mutationId, error) {
  var definition = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  return {
    type: RECEIVE_MUTATION_ERROR,
    mutationId: mutationId,
    error: error,
    definition: definition
  };
};

exports.receiveMutationError = receiveMutationError;
