// cozy-client/dist/testing/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeWrapper = exports.setupClient = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _react = _interopRequireDefault(require("react"));

var _mock = require("../mock");

var _simpsons = _interopRequireDefault(require("../testing/simpsons.json"));

var _Provider = _interopRequireDefault(require("../Provider"));

var _reactRedux = require("react-redux");

var setupClient = function setupClient() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      queries = _ref.queries;

  var client = (0, _mock.createMockClient)({
    queries: (0, _objectSpread2.default)({
      simpsons: {
        data: _simpsons.default,
        doctype: 'io.cozy.simpsons'
      },
      'simpsons/marge': {
        data: _simpsons.default.filter(function (x) {
          return x.name === 'Marge';
        }),
        doctype: 'io.cozy.simpsons',
        definition: {
          id: 'marge',
          doctype: 'io.cozy.simpsons'
        }
      },
      upperSimpsons: {
        data: _simpsons.default.map(function (x) {
          return (0, _objectSpread2.default)({}, x, {
            name: x.name.toUpperCase()
          });
        }),
        doctype: 'io.cozy.simpsons-upper'
      }
    }, queries)
  });
  client.ensureStore();
  return client;
};

exports.setupClient = setupClient;

var makeWrapper = function makeWrapper(client) {
  var Wrapper = function Wrapper(_ref2) {
    var children = _ref2.children;
    return _react.default.createElement(_reactRedux.Provider, {
      store: client.store
    }, _react.default.createElement(_Provider.default, {
      client: client
    }, children));
  };

  return Wrapper;
};

exports.makeWrapper = makeWrapper;
