// cozy-client/dist/connect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _react = _interopRequireWildcard(require("react"));

var _reactRedux = require("react-redux");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _store = require("./store");

var connect = function connect(query) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  console.warn("cozy-client's `connect` is deprecated. Please use `queryConnect` instead. For more informations about it, please see https://github.com/cozy/cozy-client/blob/master/docs/how-tos.md#how-to-connect-to-the-documents-store-declaratively-");
  return function (WrappedComponent) {
    var wrappedDisplayName = WrappedComponent.displayName || WrappedComponent.name || 'Component';

    var mapStateToProps = function mapStateToProps(state, ownProps) {
      return (0, _objectSpread2.default)({}, (0, _store.getQueryFromState)(state, ownProps.queryId));
    };

    var ConnectedWrappedComponent = (0, _reactRedux.connect)(mapStateToProps)(WrappedComponent);

    var Wrapper =
    /*#__PURE__*/
    function (_Component) {
      (0, _inherits2.default)(Wrapper, _Component);

      function Wrapper(props, context) {
        var _this;

        (0, _classCallCheck2.default)(this, Wrapper);
        _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Wrapper).call(this, props, context));
        _this.client = props.client || context.client;

        if (!_this.client) {
          throw new Error("Could not find \"client\" in either the context or props of ".concat(wrappedDisplayName));
        }

        _this.queryId = options.as || _this.client.generateId();
        _this.queryDef = typeof query === 'function' ? query(_this.client, props) : query;
        return _this;
      }

      (0, _createClass2.default)(Wrapper, [{
        key: "componentDidMount",
        value: function componentDidMount() {
          this.client.query(this.queryDef, {
            as: this.queryId
          });
        }
      }, {
        key: "render",
        value: function render() {
          return _react.default.createElement(ConnectedWrappedComponent, (0, _extends2.default)({
            queryId: this.queryId
          }, this.props));
        }
      }]);
      return Wrapper;
    }(_react.Component);

    (0, _defineProperty2.default)(Wrapper, "contextTypes", {
      client: _propTypes.default.object
    });
    Wrapper.displayName = "CozyConnect(".concat(wrappedDisplayName, ")");
    return Wrapper;
  };
};

var _default = connect;
exports.default = _default;
